src/main/java/org/dsa/iot/alarm/jdbc/JdbcProvider.java

```diff
@@ -0,0 +1,512 @@
+
+    public JdbcProvider() {
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Methods
+    ///////////////////////////////////////////////////////////////////////////
+
+    @Override public void addAlarm(final AlarmRecord arg) {
+        Connection conn = null;
+        PreparedStatement stmt = null;
+        try {
+            conn = getConnection();
+            //insert the note
+            stmt = conn.prepareStatement("insert into Alarm_Records "
+                                         + "(Uuid, "
+                                         + "SourcePath, "
+                                         + "AlarmClass, "
+                                         + "AlarmState, "
+                                         + "CreatedTime, "
+                                         + "NormalTime, "
+                                         + "AckTime, "
+                                         + "AckUser, "
+                                         + "Message, "
+                                         + "HasNotes,"
+                                         + "IsOpen) "
+                                         + "VALUES (?,?,?,?,?,?,?,?,?,?,?)");
+            stmt.setString(1, arg.getUuid().toString());
+            stmt.setString(2, arg.getSourcePath());
+            stmt.setString(3, arg.getAlarmClass().getNode().getName());
+            stmt.setString(4, AlarmState.encode(arg.getAlarmType()));
+            stmt.setTimestamp(5, new Timestamp(arg.getCreatedTime()));
+            stmt.setTimestamp(6, new Timestamp(arg.getNormalTime()));
+            stmt.setTimestamp(7, new Timestamp(arg.getAckTime()));
+            stmt.setString(8, arg.getAckUser());
+            stmt.setString(9, arg.getMessage());
+            stmt.setBoolean(10, arg.getHasNotes());
+            stmt.setBoolean(11, arg.isOpen());
+            stmt.executeUpdate();
+            conn.commit();
+        } catch (Exception x) {
+            AlarmUtil.throwRuntime(x);
+        } finally {
+            close(conn, stmt, null);
+        }
+    }
+
+    @Override protected synchronized void addNote(Note arg) {
+        Connection conn = null;
+        PreparedStatement stmt = null;
+        try {
+            conn = getConnection();
+            //insert the note
+            stmt = conn.prepareStatement("insert into Alarm_Notes "
+                                         + "(Uuid, Timestamp, User, Note) "
+                                         + "VALUES (?,?,?,?)");
+            stmt.setString(1, arg.getUUID().toString());
+            stmt.setTimestamp(2, new Timestamp(arg.getTimestamp()));
+            stmt.setString(3, arg.getUser());
+            stmt.setString(4, arg.getText());
+            stmt.executeUpdate();
+            stmt.close();
+            //update the alarm record
+            stmt = conn.prepareStatement("update Alarm_Records set HasNotes = true "
```

Comment:
see, with just a `note` field, you could compute this here, with a `count(*)` or something like that. You even have the relational aspect of it for doing it easily